MMCore/Devices/StorageInstance.cpp

```diff
@@ -0,0 +1,254 @@
+{
+   RequireInitialized(__func__);
+
+   return GetImpl()->ConfigureDimension(handle, dimension, name, meaning);
+}
+
+int StorageInstance::ConfigureCoordinate(const char* handle, int dimension, int coordinate, const char* name)
+{
+   RequireInitialized(__func__);
+
+   return GetImpl()->ConfigureCoordinate(handle, dimension, coordinate, name);
+}
+
+int StorageInstance::Close(const char* handle)
+{
+   RequireInitialized(__func__);
+
+   return GetImpl()->Close(handle);
+}
+
+int StorageInstance::Load(const char* path, std::string& handle)
+{
+   RequireInitialized(__func__);
+
+   char cHandle[MM::MaxStrLength];
+	memset(cHandle, 0, MM::MaxStrLength);
+   int ret = GetImpl()->Load(path, cHandle);
+   if (ret != DEVICE_OK)
+      return ret;
+   
+   handle = cHandle;
+   return DEVICE_OK;
+}
+
+int StorageInstance::GetShape(const char* handle, std::vector<long>& shape)
+{
+   RequireInitialized(__func__);
+   int numDim(0);
+   int ret = GetImpl()->GetNumberOfDimensions(handle, numDim);
+   if (ret != DEVICE_OK)
+      return ret;
+   int* shapeArray = new int[numDim];
+   ret = GetImpl()->GetShape(handle, shapeArray);
+   if (ret != DEVICE_OK)
+      return ret;
+   shape.clear();
+   for (int i = 0; i < numDim; i++)
+      shape.push_back(shapeArray[i]);
+
+   return DEVICE_OK;
+}
+
+int StorageInstance::GetPixelType(const char* handle, MM::StorageDataType& dataType)
+{
+   RequireInitialized(__func__);
+   int ret = GetImpl()->GetDataType(handle, dataType);
+   if (ret != DEVICE_OK)
+      return ret;
+
+   return DEVICE_OK;
+}
+
+int StorageInstance::Delete(char* handle)
+{
+   RequireInitialized(__func__);
+
+   return GetImpl()->Delete(handle);
+}
+
+int StorageInstance::List(const char* path, std::vector<std::string>& listOfDatasets)
+{
+   RequireInitialized(__func__);
+   const int maxItems(5000);
+   const int maxItemLength(1024);
+   std::vector<char*> cList(maxItems, nullptr);
+   for (auto c : cList)
+   {
+      c = new char[maxItemLength];
+      memset(c, 0, maxItemLength);
+   }
+   int ret = GetImpl()->List(path, &cList[0], maxItems, maxItemLength);
+   if (ret == DEVICE_OK)
+   {
+      listOfDatasets.clear();
+
+      for (auto c : cList)
+      {
+         if (strlen(c) == 0) break;
+         listOfDatasets.push_back(std::string(c));
+      }
+   }
+
+   for (auto c : cList) delete[] c;
+
+   return ret;
+}
+
+int StorageInstance::AddImage(const char* handle, int sizeInBytes, unsigned char* pixels, std::vector<int>& coordinates, const char* imageMeta)
+{
+   RequireInitialized(__func__);
+   return GetImpl()->AddImage(handle, sizeInBytes, pixels, &coordinates[0], (int)coordinates.size(), imageMeta);
+}
+
+int StorageInstance::AppendImage(const char* handle, int sizeInBytes, unsigned char* pixels, const char* imageMeta)
+{
+   RequireInitialized(__func__);
+   return GetImpl()->AppendImage(handle, sizeInBytes, pixels, imageMeta);
```

Comment:
MMCore should throw an error if an attempt is made to add an image to a read-only dataset (again, so that each storage device doesn't have to repeat the logic and we get a consistent error message).

Same with `AddImage` (if keeping).